chainer/variable.py

```diff
@@ -818,6 +792,142 @@ def add_cand(cand):
+    the link within an :meth:`~chainer.Link.init_scope` context.
+
+    Parameter also supports an initialization by an initializer. It can have
+    two initializers: one for the data array, and the other for the gradient
+    array. The initializer only specifies the way of filling the elements of
+    these arrays, and the shape information is specified at the initialization
+    point.
+
+    When a link that the parameter has been registered to is passed to an
+    :class:`~chainer.GradientMethod`, an update rule is set to the parameter.
+    This update rule specifies how to update the data array of the parameter
+    using its gradient array.
+
+    Args:
+        initializer (~chainer.Initializer or numpy.ndarray or cupy.ndarray):
+            Initializer of the data array. If ``shape`` is given, this
+            initializer is immediately used to initialize the data array.
+            Otherwise, if it is an array, it is immediately used as the data
+            array, and otherwise the data array is left uninitialized and will
+            be initialized by this initializer in :meth:`initialize`. It can
+            also be a scalar, in which case the data array will be filled by
+            this scalar. Note that float32 is used in this case.
+        shape (int or tuple of int or None): Shape of the parameter. If it is
+            ``None``, the initialization is deferred to the call of
+            :meth:`initialize`.
+        name (str): Name of the parameter.
+
+    Attributes:
+        initializer: Initializer of the data array. It is used for
+            initializing the data array of an uninitialized variable.
+        update_rule: :class:`~chainer.optimizer.UpdateRule` instance that
+            updates this variable as a parameter. This argument is set to
+            :attr:`update_rule`.
+
+    """
+
+    initializer = None
+    _grad_initializer = None
+    _initial_device = -1
+
+    def __init__(self, initializer=None, shape=None, name=None):
+        if initializer is None:
+            initializer = constant.NaN()
+        elif numpy.isscalar(initializer):
+            initializer = constant.Constant(initializer)
+        if shape is None:
+            if isinstance(initializer, (numpy.ndarray, cuda.ndarray)):
+                # parameter initialized by the initial array
+                super(Parameter, self).__init__(initializer, name=name)
+            else:
+                # uninitialized parameter
+                super(Parameter, self).__init__(name=name)
+                self.initializer = initializer
+                dtype = getattr(initializer, 'dtype', numpy.float32)
+                self._grad_initializer = constant.NaN(dtype)
+        else:
+            # parameter initialized with a given shape
+            if isinstance(initializer, (numpy.ndarray, cuda.ndarray)):
+                initializer = constant.Constant(initializer)
+            data = initializers.generate_array(initializer, shape, numpy)
```

Comment:
Should we notify users that parameters are initialized as `numpy.ndarray` even if `initializer` option is `cupy.ndarray`?